### PR TITLE
Allow calls to type constructors on primitives to appear in constant expressions

### PR DESCRIPTION
**Pull request by cburgdorf**
### What was wrong?

As stated in #809 one very annoying thing that arises in pretty much every attempt to write real world Fe code is that it is currently not possible to use addresses as constants.

### How was it fixed?

Hacked until it worked.

### CODE REVIEW

**Comment by Y-Nak** [comment]
Please give me a day to estimate how much this change affects the fe-v2 implementation. I guess this case would be fine.
But I want to share my feeling that I'm afraid fe-v2 implementation will never end if I have to implement all arbitrary changes in a "proper" way.

**Comment by cburgdorf** [comment]
@Y-Nak Sure, no worries. I don't mean to burden your work but this issue is a real pain when writing real world Fe code so I hope we can resolve it before the next release.

**Review by Y-Nak — APPROVED** [review]
I considered the impact and concluded it'd be fine as long as `const` is limited to top-level declarations. We need to consider it more when it is allowed to be defined inside other items like `Function` or so. So I'm happy for the PR to be merged.

**Comment by sbillig** [comment]
I think that instead of allowing primitive type constructors/casts, we should just interpret the value using the declared type, and error on values that don't fit that type.

Eg `const FOO: address = 0xfefefefe` would be allowed, but `const BAR: u8 = 1000` would be rejected.

**Comment by cburgdorf** [comment]
@sbillig That's what I initially tried to do but it wasn't obvious to me how that would work.

 `const BAR: u8 = 1000` is rejected because `1000` is parsed as a `Expr::Num(num_str)` and so we check the `num_strut in` against the permitted size directly. But in the case of `const FOO: address = 0xfefefef` the `0xfefefef` is again parsed as an `Expr::Num(num_str)` and since `address` isn't a numeric type I'm not sure how exactly we would use it to constrain the size of the numeric expression. We could say that we can *coerce* a `u256` into an address but since `address` is a smaller type than `u256` it doesn't seem right to just truncate left or right.
 
A cleaner way imho would be to add `u160` as a numeric type and make `address` an alias to it.

File: crates/analyzer/src/traversal/expressions.rs [hunk 1]
```diff
@@ -905,13 +905,13 @@ fn expr_call(
...
 
-    if context.is_in_function() {
+    if context.is_in_function() || call_type.is_primitive_type_constructor(context.db()) {
         context.add_call(func, call_type);
     } else {
         context.error(
             "calling function outside function",
             func.span,
-            "function can only be called inside function",
+            "ffunction can only be called inside function",
```
Comment (sbillig) [inline crates/analyzer/src/traversal/expressions.rs:914]: rogue f

**Comment by sbillig** [comment]
> `const BAR: u8 = 1000` is rejected because `1000` is parsed as a `Expr::Num(num_str)` and so we check the `num_str` against the permitted size directly. But in the case of `const FOO: address = 0xfefefef` the `0xfefefef` is again parsed as an `Expr::Num(num_str)` and since `address` isn't a numeric type I'm not sure how exactly we would use it to constrain the size of the numeric expression. We could say that we can _coerce_ a `u256` into an address but since `address` is a smaller type than `u256` it doesn't seem right to just truncate left or right.
> 
> A cleaner way imho would be to add `u160` as a numeric type and make `address` an alias to it.

A think Address should be a `newtype` rather than an alias, but it could certainly be defined in std in terms of u160.

For now, we can just constrain it to a u160 internally.  I think this is all doable without much trouble, by checking for `Type::Address` in `traversal::expressions::expr_num` and `Constant::from_num_str`, and checking if the string can be interpreted as a u160. This of course, would allow the use of a number literal anywhere an `address` is expected. I started along this path, but was using H160 which won't work because it expects a full 40-char hex string. I'm happy to continue, but also happy to have you do it instead :)

(Aside: perhaps non-hex address literals should be disallowed?)

**Comment by cburgdorf** [comment]
@sbillig sounds good to me!

>I'm happy to continue

Yes, please take it, I'm happy to review it then.

File: crates/analyzer/src/traversal/expressions.rs (hunk 1 above)
Comment (cburgdorf) [reply crates/analyzer/src/traversal/expressions.rs:914]: oops

**Pull request closed without merging**